### ISSUE TITLE
fp16 support checklist

### ISSUE THREAD

**Issue by gchanan** [torch/cunn#347]
- [x] Abs
- [x] ELU
- [x] HardTanh
- [x] LeakyReLU
- [x] LogSigmoid
- [x] LogSoftMax
- [x] PReLU
- [x] RReLU
- [x] Sigmoid
- [x] SoftMax
- [x] SoftPlus
- [x] SoftShrink
- [x] Square
- [x] Sqrt
- [x] Tanh
- [x] Threshold
- [x] AbsCriterion
- [x] BatchNormalization
- [x] BCECriterion
- [x] ClassNLLCriterion
- [x] DistKLDivCriterion
- [x] L1Cost
- [x] LookupTable
- [x] MarginCriterion
- [x] MSECriterion
- [x] MultiLabelMarginCriterion
- [x] MultiMarginCriterion
- [x] SmoothL1Criterion
- [x] SoftMarginCriterion
- [x] SparseLinear
- [x] SpatialAdaptiveMaxPooling
- [x] SpatialAveragePooling
- [x] SpatialClassNLLCriterion
- [x] SpatialConvolutionLocal
- [x] SpatialConvolutionMM
- [x] SpatialCrossMapLRN
- [x] SpatialDilatedConvolution
- [x] SpatialDilatedMaxPooling
- [x] SpatialFractionalMaxPooling
- [x] SpatialFullConvolution
- [x] SpatialMaxPooling
- [x] SpatialMaxUnpooling
- [x] SpatialReflectionPadding
- [x] SpatialReplicationPadding
- [x] SpatialSubSampling
- [x] SpatialUpSamplingBilinear
- [x] SpatialUpSamplingNearest
- [x] TemporalConvolution
- [x] TemporalMaxPooling
- [x] VolumetricAveragePooling
- [x] VolumetricConvolution
- [x] VolumetricDilatedConvolution
- [x] VolumetricDilatedMaxPooling
- [x] VolumetricFullConvolution
- [x] VolumetricMaxPooling
- [x] VolumetricMaxUnpooling
- [x] VolumetricReplicationPadding

**Comment by ngimel** [comment]
How do you handle computation type being potentially different from I/O type (e.g., I/O in fp16, computation in fp32)? For bandwidth-bound operations, it makes sense to always perform computation in fp32, even on Pascal cards with fp16 I/O  tensors for two reasons
1) Even on P100, there is no to negligible performance benefit for using fp16 math for bandwidth-bound operations, yet accuracy degradation can be noticeable (we had to turn fp16 math off in cutorch).
2) On pascal cards other than P100, fp16 math throughput is very low, so there will be performance penalty.

**Comment by soumith** [comment]
@ngimel for bandwidth-bound operations which need high-precision math, we are doing the operations in float / double.

For the rest, we use HALF instructions if they are available, and fallback to float32 otherwise.

This needs to be rethought of course, I think a global compile-time flag exists now, but making it a global runtime flag might be worth thinking.

For current progress, have a look at: https://github.com/torch/cunn/pull/346

**Comment by borisfom** [comment]
@soumith , @gchanan : I was thinking, Torch could do better than global runtime flag. 
For maximum flexibility, we can do 2 distinct types: NativeHalf and PseudoHalf, and Half would be a 'typedef' to one of them controlled by runtime  switch.  I have started to work on a POC in cutorch: replacing #ifdefs with distinct specializations for NativeHalf and PseudoHalf. The code becomes a lot cleaner! I also believe Native and Pseudo should be exposed to the user so, for example, in CUDNN we can use NativeHalf for forward path and Pseudo for backward :we are using some ugly flags in Caffe to acieve that but Torch can express it much more naturally. 
@soumith: please let me know if that sounds right for you and if I should continue on it.

**Comment by shuzi** [comment]
@gchanan 
nn.Linear is not in the list.  Does it mean that nn.Linear does not support FP16?
since I can pass a FP16 tensor to nn.Linear , it does not report any error

**Comment by fmassa** [comment]
@shuzi nn.Linear supports FP16. It's not in the list because it is implemented using GEMM (matrix multiply), which has FP16 support
